ir/src/transition_constraints/graph.rs

```diff
@@ -0,0 +1,146 @@
+use crate::error::SemanticError;
+
+use super::TraceColumns;
+use parser::ast::{self, Expr, Identifier};
+
+// ALGEBRAIC GRAPH
+// ================================================================================================
+
+/// The AlgebraicGraph is a directed acyclic graph used to represent transition constraints. To
+/// store it compactly, it is represented as a vector of nodes where each node references other
+/// nodes by their index in the vector.
+///
+/// Within the graph, constraint expressions can overlap and share subgraphs, since new expressions
+/// reuse matching existing nodes when they are added, rather than creating new nodes.
+///
+/// - Leaf nodes (with no outgoing edges) are constants or references to trace cells (i.e. column 0
+///   in the current row or column 5 in the next row).
+/// - Tip nodes with no incoming edges (no parent nodes) always represent constraints, although they
+///   do not necessarily represent all constraints. There could be constraints which are also
+///   subgraphs of other constraints.
+#[derive(Default, Debug)]
+pub struct AlgebraicGraph {
+    /// All nodes in the graph.
+    nodes: Vec<Node>,
+}
+
+impl AlgebraicGraph {
+    // --- PUBLIC ACCESSORS -----------------------------------------------------------------------
+
+    /// Returns the node with the specified index.
+    pub fn node(&self, index: &NodeIndex) -> &Node {
+        &self.nodes[index.0]
+    }
+
+    /// Returns the degree of the subgraph which has the specified node as its tip.
+    pub fn degree(&self, index: &NodeIndex) -> u8 {
```

Comment:
This is fine for now, but in the future (once we start working with periodic columns), the return type would need to be something like a `DegreeDescriptor` struct from Winterfell.